internal/cephfs/controllerserver.go

```diff
@@ -36,25 +40,82 @@ type ControllerServer struct {
 	// A map storing all volumes with ongoing operations so that additional operations
 	// for that same volume (as defined by VolumeID/volume name) return an Aborted error
 	VolumeLocks *util.VolumeLocks
+
+	// A map storing all snapshots with ongoing operations so that additional operations
+	// for that same snapshot (as defined by SnapshotID/snapshot name) return an Aborted error
+	SnapshotLocks *util.VolumeLocks
 }
 
 // createBackingVolume creates the backing subvolume and on any error cleans up any created entities.
-func (cs *ControllerServer) createBackingVolume(ctx context.Context, volOptions *volumeOptions, vID *volumeIdentifier, secret map[string]string) error {
-	cr, err := util.NewAdminCredentials(secret)
-	if err != nil {
-		return status.Error(codes.InvalidArgument, err.Error())
+func (cs *ControllerServer) createBackingVolume(
+	ctx context.Context,
+	volOptions *volumeOptions,
+	vID *volumeIdentifier,
+	parentVolOpt *volumeOptions,
+	pvID *volumeIdentifier,
+	sID *snapshotIdentifier,
+	cr *util.Credentials) error {
+	var err error
+	if sID != nil {
+		err = createCloneFromSnapshot(ctx, parentVolOpt, volOptions, pvID, vID, sID, cr)
+		var ecip ErrCloneInProgress
+		if !errors.As(err, &ecip) {
+			return status.Error(codes.Internal, err.Error())
+		}
+		return err
+	}
+	if parentVolOpt != nil {
+		err = createCloneFromSubvolume(ctx, volumeID(pvID.FsSubvolName), volumeID(vID.FsSubvolName), volOptions, parentVolOpt, cr)
+		if err != nil {
```

Comment:
The function `createCloneFromSubvolume` can return `ErrCloneInProgress`, it looks like we would not process that here and instead return an internal error back to the caller?